README.txt
==== Ubuntu build/install====
1. sudo apt-get install build-dep qrfcview
2. cd qRFCView-deb-package; dpkg-buildpackage -rfakeroot -uc -b
3. cd ..; dpkg -i qrfcview_0.62-5.3_amd64.deb

==== qRFCView ====

== What is qRFCView? ==

 qRFCView is a viewer for IETF RFC.
 
 Advantages:
  - automatic table of content, with direct opening of section;
  - handling of RFC internal cross-references;
  - automatic downloading of a referenced RFC from the IETF web site
    on a simple click;
  - caching of RFC in a local directory;
  - tab-browsing of RFC;
  - searching.
  
 It is available for Linux and Windows operating system (and
 potentially MacOS X).

== How to install it? ==

 qRFCView requires the Trolltech Qt4 library that is available under MAC OS, 
 Windows and Linux platforms. Trolltech's dual license policy provides versions 
 of Qt under the terms of the GNU General Public License (GPL). It can be
 downloaded from http://www.trolltech.com/products/qt/downloads.
 
 * Win32 building (WinNT/Win2k/WinXP)
 Makefile generation: qmake -win32 rfcview.pro
 qRFCView binary build: make
 Optionally, you can build a Windows installer
 
 * Unix building
 qmake -unix rfcview.pro
 qRFCView binary build: make
  
 * Win32 binary
 The qRFCView binary can be directly installed by using the provided Win32 
 installer file. The following DLLs will be added to the windows system
 directory:
	- mingwm10.dll
    - QtCore4.dll
    - QtDesignerComponents4.dll
    - QtGui4.dll
    - QtNetwork4.dll

== How to use it? ==

Just start it. Use "File>Load RFC..." to load a RFC by its number.

Double-click on a table of content item to open the corresponding
section. Internal links and referenced RFC are active hyperlinks,
underlined in blue.

One can set the default directories where RFC are saved using
"Edit>Set Directories...".

== License ==

qRFCView is free software, licensed under GNU GPL (see LICENSE file).

== Contact ==
qRFCView has been developped at Mitsubishi Electric, ITE-TCL, France.
qRFCView developer is Romain ROLLET <rfcview@gmail.com>.


== Known bugs ==

#1 Under Windows, qRFCView sometimes doesn't close when exiting Windows while
   qRFCView is still open.


debian/changelog
qrfcview (0.62-5.3) unstable; urgency=low

  * Non-maintainer update.
  * d/p/04_find_dialog_focus.patch: the QComboBox get focus, when find dialog
    opened.
  * d/p/05_optimize_tabwidget.patch: Tabwidget can be movable.

 -- Ji Tao <jt6562@gmail.com>  Fri, 9 Sep 2016 10:58:03 +0800

qrfcview (0.62-5.2) unstable; urgency=low

  * Non-maintainer upload.
  * d/p/03-max_rfc_to_99999.patch: allow RFC numbers larger than 5000
    (Closes: #512734)

 -- Laurent Bigonville <bigon@debian.org>  Tue, 19 Feb 2013 15:45:19 +0100

qrfcview (0.62-5.1) unstable; urgency=low

  * Non-maintainer upload.
  * Fix build with gcc-4.4 using patch from Martin Michlmayr (Closes:
    #505341)

 -- Christoph Egger <christoph@debian.org>  Fri, 25 Dec 2009 22:20:17 +0100

qrfcview (0.62-5) unstable; urgency=low

  * debian/control
    * Updated Standards-Version to 3.7.3
    * Added Homepage: field
  * debian/menu: use now menu-2 format
  * debian/copyright: adapted to the machine-interpretable proposal in
    http://wiki.debian.org/Proposals/CopyrightFormat
  * Replaced dpatch with quilt
  * debian/rules: improved clean target thanks to Daniel Baumann

 -- Frederic Daniel Luc Lehobey <Frederic@Lehobey.net>  Wed, 23 Apr 2008 18:10:48 +0200

qrfcview (0.62-4) unstable; urgency=low

  * Changed maintainer address
  * Fixed a typo in manpage
  * Added a debian/watch file
  * Pushed debian/compat to 5
  * debian/control: removed redundant build-dependency on libglib2.0-dev
    and pushed debhelper version
  * Being lintian clean again
    * Updated menu entry to fit new policy
    * Updated Freedesktop menu entry to menu-spec 1.0
    * debian/rules: minor fixes

 -- Frederic Daniel Luc Lehobey <Frederic@Lehobey.net>  Tue, 21 Aug 2007 12:51:00 +0200

qrfcview (0.62-3) unstable; urgency=low

  * debian/copyright: clarified icons licenses
  * debian/rules, control and patches: introduced dpatch use for
    01-removal-of-spurious-debug-output.dpatch
  * debian/control: added build-dependency on libglib2.0-dev (and dpatch)

 -- Frederic Daniel Luc Lehobey <Frederic.Lehobey@free.fr>  Sat, 14 Oct 2006 23:41:51 +0200

qrfcview (0.62-2) unstable; urgency=low

  * debian/rules: renamed generated icons to qrfcview instead of rfcview thanks
    to Daniel Baumann <daniel@debian.org>
  * debian/control: added Suggests: doc-rfc

 -- Frederic Daniel Luc Lehobey <Frederic.Lehobey@free.fr>  Fri, 28 Jul 2006 00:33:57 +0200

qrfcview (0.62-1) unstable; urgency=low

  * Initial release (Closes: #378587)
  * debian/control: fixed duplicate dependencies thanks to Ted Percival
    <ted@midg3t.net>
  * debian/control: Build-conflicts with qt3-dev-tools thanks to andeus
  * debian/README.Debian: clarified doc-rfc usage thanks to Neil Williams
    <linux@codehelp.co.uk>
  * Much improved package quality thanks to Daniel Baumann <daniel@debian.org>
    careful mentoring, patch and advice

 -- Frederic Daniel Luc Lehobey <Frederic.Lehobey@free.fr>  Thu, 27 Jul 2006 10:57:03 +0200


src/mdichild.h
/****************************************************************************
  
    qRFCView, A smart IETF RFC viewer based on the Qt4 library.
    Copyright (C) 2005 Mitsubishi Electric ITE-TCL, R. Rollet (rollet@tcl.ite.mee.com)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*********************************************************************************/



    #ifndef MDICHILD_H
    #define MDICHILD_H

    #include <QSplitter>

    #include <stdint.h>

#define PARSER_NORMALTEXT 0
#define PARSER_BIBTEXT 1
#define PARSER_HEADER 2
#define PARSER_TRAILER 3

    class QTreeView;
    class QRFCEditor;
    class CTitleModel;
    class QModelIndex;
    class QFile;
    class QUrl;
    class QPrinter;
        
    class MdiChild : public QSplitter
    {
        Q_OBJECT

    public:
        MdiChild(QWidget *pParent=0);
        ~MdiChild();
        bool loadFile(const QString &fileName);
        QString userFriendlyCurrentFile();
        QString currentFile() { return curFile; }
        void copy();
        bool hasSelection();
        void setCurrentFont(const QFont &);
        bool FindText(QString &qTextToFind, uint32_t iOptionFlags);
        void Print(QPrinter *pPrinter, bool all, int fromPage, int toPage);
        uint32_t GetNbPages() {return m_iNbPages;}
                
    protected:
        void closeEvent(QCloseEvent *event);

    private:
        void setCurrentFile(const QString &fileName);
        QString strippedName(const QString &fullFileName);
        bool ParseRFCText(const QFile &qFile, QString &qOutput);

    public slots:
        void goToTitle(const QModelIndex &);
        void goToAnchor(const QUrl &qURL);
    
    public:
      QRFCEditor *m_pTextEdit;     
    private:
      QTreeView *m_pTreeView;              
      CTitleModel *m_pTitleModel; 
      QString m_qFileName;
      QString curFile;
      bool isUntitled;
      uint32_t m_iNbPages;
      
    };

    #endif


src/qrfceditor.h
/****************************************************************************
  
    qRFCView, A smart IETF RFC viewer based on the Qt4 library.
    Copyright (C) 2005 Mitsubishi Electric ITE-TCL, R. Rollet (rollet@tcl.ite.mee.com)
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*********************************************************************************/

#ifndef QRFCEDITOR_H
#define QRFCEDITOR_H

#include <qtextbrowser.h>

#include <stdint.h>

class QStringList;

/**
@author Romain Rollet
*/
class QRFCEditor : public QTextBrowser
{
Q_OBJECT
public:
    QRFCEditor(QWidget *parent = 0);

    ~QRFCEditor();
  
    void setSource ( const QUrl & name );
    void scrollToAnchor2 ( const QString & name );
    bool isBackwardAvailable ();
    bool isForwardAvailable ();
    
signals:
    void RFCReq(uint32_t); // Signal emitted when a new RFC is required    
public slots:
    void backward ();
    void forward ();
    
private:
    QList<int> m_qPositionPath;
    int m_iCurrentPositionIdx;
};

#endif


src/mainwindow.cpp
/****************************************************************************

    qRFCView, A smart IETF RFC viewer based on the Qt4 library.
    Copyright (C) 2005 Mitsubishi Electric ITE-TCL, R. Rollet (rollet@tcl.ite.mee.com)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*********************************************************************************/


#include <QtGui>
#include <QFontDialog>
#include <QHttp>
#include <QPrinter>
#include <QLibrary>

#include "rfcloader.h"
#include "mainwindow.h"
#include "mdichild.h"
#include "cdialogsetdirectory.h"
#include "cdialogfind.h"
#include "qrfceditor.h"
#include "cprintdialog.h"

MainWindow::MainWindow():m_qLastOpenDir(".")
{
    m_qTabWidget = new QTabWidget(this);

    m_qTabWidget->setMovable(true);

    setCentralWidget(m_qTabWidget);

    connect(m_qTabWidget, SIGNAL(currentChanged(int)),
            this, SLOT(updateMenus()));
            /*
    windowMapper = new QSignalMapper(this);
    connect(windowMapper, SIGNAL(mapped(QWidget *)),
            workspace, SLOT(setActiveWindow(QWidget *)));
            */
    readSettings();
    QUrl qURL(m_qRFCURL);
    m_pRFCLoader=new QRFCLoader(this);
    connect(m_pRFCLoader, SIGNAL(done(QString)),this, SLOT(RFCReady(QString)));
    connect(m_pRFCLoader, SIGNAL(start(QString)),this, SLOT(RFCStart(QString)));
    m_pRFCLoader->SetDownloadURL(qURL);
    m_pRFCLoader->SetDirectories(m_qDirectoryList, m_iDefaultDirectory);

    createActions();
    createMenus();
    createToolBars();
    createStatusBar();
    updateMenus();
    //m_pDialogFind=NULL;
    m_pDialogFind=new CDialogFind(  this);
    connect(m_pDialogFind, SIGNAL(findnext()), this, SLOT(findnext()));
    setWindowTitle(tr("qRFCView"));
    //m_pRFCLoader->GetFile(794);

}

void MainWindow::closeEvent(QCloseEvent *event)
{
    //workspace->closeAllWindows();
    /*
    if (activeMdiChild()) {
        event->ignore();
    } else {
        writeSettings();
        event->accept();
    }*/
    writeSettings();
    event->accept();
}


void MainWindow::open()
{
    QString fileName = QFileDialog::getOpenFileName(this, QString(), m_qLastOpenDir.absolutePath() );
    QFileInfo qFileInfo(fileName);

    if (!fileName.isEmpty()) {
        MdiChild *existing = findMdiChild(fileName);
        if (existing) {
            m_qTabWidget->setCurrentWidget(existing);
            return;
        }

        MdiChild *child = createMdiChild(qFileInfo.fileName());
        child->setCurrentFont(m_qFont);
        if (child->loadFile(fileName)) {
            statusBar()->showMessage(tr("File loaded"), 2000);
            child->show();
            m_qLastOpenDir=qFileInfo.dir();
        } else {
            child->close();
        }
    }
}

void MainWindow::close()
{
  MdiChild *pMdiChild=activeMdiChild();

  if (pMdiChild)
  {
    m_qTabWidget->removeTab(m_qTabWidget->currentIndex());
    delete pMdiChild;
  }
  updateMenus();
}

void MainWindow::getrfc()
{
  // Load a RFC number
  bool bOK;
  int iRFCNum = QInputDialog::getInteger(this, tr("Please enter a RFC number"),
                                             tr("RFC#:"), 0, 1, 99999, 1, &bOK);
  if (bOK)
    RFCLoad( iRFCNum );
}


void MainWindow::copy()
{
    activeMdiChild()->copy();
}

void MainWindow::findOpen()
{
    m_pDialogFind->show();
    m_pDialogFind->raise();
    m_pDialogFind->activateWindow();
}

void MainWindow::findnext()
{
  bool bResult;
  uint32_t iOptionFlags;

  QString qTextToFind=m_pDialogFind->GetTextToFind();
  iOptionFlags=m_pDialogFind->GetOptionFlags();
  MdiChild *pMdiChild=activeMdiChild();
  if (pMdiChild && !qTextToFind.isEmpty() )
  {
    bResult=pMdiChild->FindText(qTextToFind, iOptionFlags );
    if (!bResult)
      statusBar()->showMessage(tr("No more occurence"), 3000);
    else
      statusBar()->clearMessage();
    iOptionFlags|=FIND_OPTIONSFLAG_CURSOR;
    m_pDialogFind->SetOptionFlags(iOptionFlags);
  }
  updateMenus();
}

void MainWindow::findprev()
{
  bool bResult;
  uint32_t iOptionFlags;

  QString qTextToFind=m_pDialogFind->GetTextToFind();
  iOptionFlags=m_pDialogFind->GetOptionFlags();
  MdiChild *pMdiChild=activeMdiChild();
  if (pMdiChild && !qTextToFind.isEmpty() )
  {
    iOptionFlags|=FIND_OPTIONSFLAG_CURSOR|FIND_OPTIONSFLAG_BACKWARD;
    bResult=pMdiChild->FindText(qTextToFind, iOptionFlags );
    if (!bResult)
      statusBar()->showMessage(tr("No more occurence"), 3000);
    else
      statusBar()->clearMessage();

  }
}

void MainWindow::forward()
{
  MdiChild *pMdiChild=activeMdiChild();
  if (pMdiChild)
    pMdiChild->m_pTextEdit->forward();
}


void MainWindow::backward()
{
  MdiChild *pMdiChild=activeMdiChild();
  if (pMdiChild)
    pMdiChild->m_pTextEdit->backward();
}

void MainWindow::setDirectories()
{
    CDialogSetDirectory *pDialog=new CDialogSetDirectory(&m_qDirectoryList, m_iDefaultDirectory, m_qRFCURL, this);
    if (pDialog->exec()==QDialog::Accepted)
    {
      m_iDefaultDirectory=pDialog->GetDirectoryList(&m_qDirectoryList);
      m_qRFCURL=pDialog->GetRFCURL();
      QUrl qURL(m_qRFCURL);
      m_pRFCLoader->SetDownloadURL(qURL);
      m_pRFCLoader->SetDirectories(m_qDirectoryList, m_iDefaultDirectory);
    }
    delete pDialog;

}

void MainWindow::setFont()
{
    bool bOK;
    MdiChild *pMDIChild;
    QFont qFont=QFontDialog::getFont(&bOK, m_qFont, this );
    m_qFont=qFont;
    if (bOK)
    {
      for (int i=0;i<m_qTabWidget->count();i++)
      {
        pMDIChild = qobject_cast<MdiChild *>(m_qTabWidget->widget(i));
        pMDIChild->setCurrentFont(m_qFont);
      }
    }
}

void MainWindow::print()
{
  MdiChild *pMdiChild=activeMdiChild();
  if (pMdiChild)
  {
    QPrinter printer(QPrinter::PrinterResolution);

    printer.setFullPage(true);
    printer.setNumCopies(2);
    QPrintDialog dlg(&printer, this);
    //dlg.setPrintRange(QAbstractPrintDialog::PageRange);
    dlg.setMinMax(1, pMdiChild->GetNbPages() );
    //dlg.setFromTo(1, 2);
    //dlg.setEnabledOptions(QAbstractPrintDialog::PrintPageRange);

    if (dlg.exec() == QDialog::Accepted) {
        pMdiChild->Print(&printer, dlg.printRange()==QAbstractPrintDialog::AllPages, dlg.fromPage(), dlg.toPage() );
        //pMdiChild->Print(&printer, false, 4,6);
    }

  }
}

void MainWindow::about()
{
    QMessageBox::about(this, tr("About qRFCView"),
            tr("<b>qRFCView v0.62</b><br> A smart RFC Viewer using the Qt4 library.<br><br> Mitsubishi Electric, 2005"));
}

void MainWindow::updateMenus()
{
/*
    bool hasMdiChild = (activeMdiChild() != 0);

    closeAct->setEnabled(hasMdiChild);
    closeAllAct->setEnabled(hasMdiChild);
    tileAct->setEnabled(hasMdiChild);
    cascadeAct->setEnabled(hasMdiChild);
    nextAct->setEnabled(hasMdiChild);
    previousAct->setEnabled(hasMdiChild);
    separatorAct->setVisible(hasMdiChild);
*/
    bool hasSelection = (activeMdiChild() &&
                          activeMdiChild()->hasSelection());
    copyAct->setEnabled(hasSelection);
    findAct->setEnabled( activeMdiChild()!=NULL);
    printAct->setEnabled( activeMdiChild()!=NULL);
    findnextAct->setEnabled( activeMdiChild()!=NULL && (!m_pDialogFind->GetTextToFind().isEmpty()) );
    findprevAct->setEnabled( activeMdiChild()!=NULL && (!m_pDialogFind->GetTextToFind().isEmpty()) );
    forwardAct->setEnabled( activeMdiChild()!=NULL && activeMdiChild()->m_pTextEdit->isForwardAvailable() );
    backwardAct->setEnabled( activeMdiChild()!=NULL && activeMdiChild()->m_pTextEdit->isBackwardAvailable() );
    closeAct->setEnabled(activeMdiChild()!=NULL);
}

void MainWindow::updateWindowMenu()
{
    /*
    windowMenu->clear();
    windowMenu->addAction(closeAct);
    windowMenu->addAction(closeAllAct);
    windowMenu->addSeparator();
    windowMenu->addAction(tileAct);
    windowMenu->addAction(cascadeAct);
    windowMenu->addSeparator();
    windowMenu->addAction(nextAct);
    windowMenu->addAction(previousAct);
    windowMenu->addAction(separatorAct);

    QList<QWidget *> windows = workspace->windowList();
    separatorAct->setVisible(!windows.isEmpty());

    for (int i = 0; i < windows.size(); ++i) {
        MdiChild *child = qobject_cast<MdiChild *>(windows.at(i));

        QString text;
        if (i < 9) {
            text = tr("&%1. %2").arg(i + 1)
                                .arg(child->userFriendlyCurrentFile());
        } else {
            text = tr("%1. %2").arg(i + 1)
                                .arg(child->userFriendlyCurrentFile());
        }
        QAction *action  = windowMenu->addAction(text);
        action->setCheckable(true);
        action ->setChecked(child == activeMdiChild());
        connect(action, SIGNAL(triggered()), windowMapper, SLOT(map()));
        windowMapper->setMapping(action, child);
    }*/
}

MdiChild *MainWindow::createMdiChild(const QString &qTitle)
{
    MdiChild *child = new MdiChild();
    m_qTabWidget->addTab(child, qTitle);

    connect(child, SIGNAL(copyAvailable(bool)),
            copyAct, SLOT(setEnabled(bool)));
    connect(child->m_pTextEdit, SIGNAL(RFCReq( uint32_t )), this, SLOT(RFCLoad( uint32_t  )) );
    connect(child->m_pTextEdit, SIGNAL(forwardAvailable(bool)), forwardAct, SLOT(setEnabled(bool)) );
    connect(child->m_pTextEdit, SIGNAL(backwardAvailable(bool)), backwardAct, SLOT(setEnabled(bool)) );
    m_qTabWidget->setCurrentWidget(child);
    return child;
}

void MainWindow::createActions()
{
    openAct = new QAction(QIcon(":/images/open.png"), tr("&Open..."), this);
    openAct->setShortcut(tr("Ctrl+O"));
    openAct->setStatusTip(tr("Open an existing file"));
    connect(openAct, SIGNAL(triggered()), this, SLOT(open()));

    loadAct = new QAction(QIcon(":/images/load.png"), tr("&Load RFC..."), this);
    loadAct->setShortcut(tr("Ctrl+L"));
    loadAct->setStatusTip(tr("Load a RFC from its number"));
    connect(loadAct, SIGNAL(triggered()), this, SLOT(getrfc()));

    closeAct = new QAction(QIcon(":/images/close.png"),tr("Cl&ose"), this);
    closeAct->setShortcut(tr("Ctrl+F4"));
    closeAct->setStatusTip(tr("Close the active window"));
    connect(closeAct, SIGNAL(triggered()), this, SLOT(close()));


    printAct = new QAction(QIcon(":/images/print.png"), tr("&Print..."), this);
    connect(printAct, SIGNAL(triggered()), this, SLOT(print()));


    exitAct = new QAction(tr("E&xit"), this);
    exitAct->setShortcut(tr("Ctrl+Q"));
    exitAct->setStatusTip(tr("Exit the application"));
    connect(exitAct, SIGNAL(triggered()), qApp, SLOT(closeAllWindows()));

    copyAct = new QAction(QIcon(":/images/copy.png"), tr("&Copy"), this);
    copyAct->setShortcut(tr("Ctrl+C"));
    copyAct->setStatusTip(tr("Copy the current selection's contents to the "
                              "clipboard"));
    connect(copyAct, SIGNAL(triggered()), this, SLOT(copy()));

    setFontAct = new QAction(QIcon(), tr("&Set Font..."), this);
    connect(setFontAct, SIGNAL(triggered()), this, SLOT(setFont()));
    setDirectoriesAct = new QAction(QIcon(), tr("&Set Directories..."), this);
    connect(setDirectoriesAct, SIGNAL(triggered()), this, SLOT(setDirectories()));

    forwardAct = new QAction(QIcon(":/images/forward.png"), tr("&Forward"), this);
    connect(forwardAct, SIGNAL(triggered()), this, SLOT(forward()));
    backwardAct = new QAction(QIcon(":/images/backward.png"), tr("&Backward"), this);
    connect(backwardAct, SIGNAL(triggered()),this, SLOT(backward()));
    findAct = new QAction(QIcon(":/images/find.png"), tr("&Find..."), this);
    findAct->setShortcut(tr("Ctrl+F"));
    findAct->setStatusTip(tr("Find a string in the currently opened file"));
    connect(findAct, SIGNAL(triggered()), this, SLOT(findOpen()));
    findnextAct = new QAction(QIcon(), tr("&Find Next"), this);
    findnextAct->setShortcut(tr("F3"));
    connect(findnextAct, SIGNAL(triggered()), this, SLOT(findnext()));
    findprevAct = new QAction(QIcon(), tr("&Find Previous"), this);
    findprevAct->setShortcut(tr("Shift+F3"));
    connect(findprevAct, SIGNAL(triggered()), this, SLOT(findprev()));

    /*


    closeAllAct = new QAction(tr("Close &All"), this);
    closeAllAct->setStatusTip(tr("Close all the windows"));
    connect(closeAllAct, SIGNAL(triggered()),
            workspace, SLOT(closeAllWindows()));

    tileAct = new QAction(tr("&Tile"), this);
    tileAct->setStatusTip(tr("Tile the windows"));
    connect(tileAct, SIGNAL(triggered()), workspace, SLOT(tile()));

    cascadeAct = new QAction(tr("&Cascade"), this);
    cascadeAct->setStatusTip(tr("Cascade the windows"));
    connect(cascadeAct, SIGNAL(triggered()), workspace, SLOT(cascade()));

    nextAct = new QAction(tr("Ne&xt"), this);
    nextAct->setShortcut(tr("Ctrl+F6"));
    nextAct->setStatusTip(tr("Move the focus to the next window"));
    connect(nextAct, SIGNAL(triggered()),
            workspace, SLOT(activateNextWindow()));

    previousAct = new QAction(tr("Pre&vious"), this);
    previousAct->setShortcut(tr("Ctrl+Shift+F6"));
    previousAct->setStatusTip(tr("Move the focus to the previous "
                                  "window"));
    connect(previousAct, SIGNAL(triggered()),
            workspace, SLOT(activatePreviousWindow()));

    separatorAct = new QAction(this);
    separatorAct->setSeparator(true);
*/

    aboutAct = new QAction(tr("&About"), this);
    aboutAct->setStatusTip(tr("Show the application's About box"));
    connect(aboutAct, SIGNAL(triggered()), this, SLOT(about()));

    aboutQtAct = new QAction(tr("About &Qt"), this);
    aboutQtAct->setStatusTip(tr("Show the Qt library's About box"));
    connect(aboutQtAct, SIGNAL(triggered()), qApp, SLOT(aboutQt()));
}

void MainWindow::createMenus()
{
    fileMenu = menuBar()->addMenu(tr("&File"));
    fileMenu->addAction(openAct);
    fileMenu->addAction(loadAct);
    fileMenu->addAction(closeAct);
    fileMenu->addAction(printAct);
    fileMenu->addSeparator();
    fileMenu->addAction(exitAct);

    editMenu = menuBar()->addMenu(tr("&Edit"));
    editMenu->addAction(backwardAct);
    editMenu->addAction(forwardAct);
    editMenu->addAction(copyAct);
    editMenu->addAction(findAct);
    editMenu->addAction(findprevAct);
    editMenu->addAction(findnextAct);
    editMenu->addSeparator();
    editMenu->addAction(setFontAct);
    editMenu->addAction(setDirectoriesAct);

    //windowMenu = menuBar()->addMenu(tr("&Window"));
    //connect(windowMenu, SIGNAL(aboutToShow()), this, SLOT(updateWindowMenu()));

    //menuBar()->addSeparator();

    helpMenu = menuBar()->addMenu(tr("&Help"));
    helpMenu->addAction(aboutAct);
    helpMenu->addAction(aboutQtAct);
}

void MainWindow::createToolBars()
{
    fileToolBar = addToolBar(tr("File"));
    fileToolBar->addAction(openAct);
    fileToolBar->addAction(loadAct);
    fileToolBar->addAction(closeAct);
    fileToolBar->addAction(printAct);

    editToolBar = addToolBar(tr("Edit"));
    editToolBar->addAction(copyAct);
    editToolBar->addAction(findAct);
    editToolBar->addAction(backwardAct);
    editToolBar->addAction(forwardAct);
}

void MainWindow::createStatusBar()
{
    m_pProgressBar=new QProgressBar();
    m_pProgressBar->setRange(0,100);
    m_pProgressBar->setValue(50);
    connect(m_pRFCLoader->GetQHttp(), SIGNAL(dataReadProgress(int,int)), this, SLOT( updateRFCProgress(int,int)));
    statusBar()->showMessage(tr("Ready"));
    //statusBar()->addWidget(m_pProgressBar, 0);
    //qProgressBar.show();
}

void MainWindow::readSettings()
{
    int i;
    QSettings settings("MELCO", "qRFCView");
    QPoint pos = settings.value("pos", QPoint(200, 200)).toPoint();
    QSize size = settings.value("size", QSize(400, 400)).toSize();
    m_qFont.setFamily(settings.value("Font_family", m_qFont.family()).toString() );
    m_qFont.setPointSize(settings.value("Font_size",   m_qFont.pointSize()).toInt() );
    m_qFont.setWeight(settings.value("Font_weight", m_qFont.weight()).toInt());
    m_qFont.setItalic(settings.value("Font_italic", m_qFont.italic()).toBool());
    m_qDirectoryList=settings.value("Directories", QStringList(QString(".") )).toStringList();
    m_iDefaultDirectory=settings.value("DefaultDirectory", 0).toInt();
    m_qRFCURL=settings.value("RFC_URL", QString("http://www.ietf.org/rfc/") ).toString();

    move(pos);
    resize(size);
    for (i=0;i<m_qDirectoryList.count();i++)
      if (m_qDirectoryList[i].isEmpty())
        m_qDirectoryList.removeAt(i);
    if (m_qDirectoryList.count()==0)
    {
      m_qDirectoryList=QStringList(QString("."));
      m_iDefaultDirectory=0;
    }
    // Check DefaultDirectory index
    if (m_iDefaultDirectory>=m_qDirectoryList.size() )
      m_iDefaultDirectory=0;
    //qDebug() << m_qDirectoryList.count();
    m_qLastOpenDir=QDir( settings.value("LastOpenedDirectory", m_qDirectoryList[m_iDefaultDirectory]).toString() );
    //m_qLastOpenDir=QDir(m_qDirectoryList[m_iDefaultDirectory]);
}

void MainWindow::writeSettings()
{
    QSettings settings("MELCO", "qRFCView");
    settings.setValue("pos", pos());
    settings.setValue("size", size());
    settings.setValue("Font_family", m_qFont.family());
    settings.setValue("Font_size",   m_qFont.pointSize());
    settings.setValue("Font_weight", m_qFont.weight());
    settings.setValue("Font_italic", m_qFont.italic());
    settings.setValue("Directories", m_qDirectoryList);
    settings.setValue("DefaultDirectory", m_iDefaultDirectory);
    settings.setValue("RFC_URL", m_qRFCURL );
    settings.setValue("LastOpenedDirectory", m_qLastOpenDir.absolutePath() );
}

MdiChild *MainWindow::activeMdiChild()
{

    return qobject_cast<MdiChild *>(m_qTabWidget->currentWidget());
}

MdiChild *MainWindow::findMdiChild(const QString &fileName)
{
    QString canonicalFilePath = QFileInfo(fileName).canonicalFilePath();

    for (int i=0;i<m_qTabWidget->count();i++)
    {
       MdiChild *mdiChild = qobject_cast<MdiChild *>(m_qTabWidget->widget(i));
        if (mdiChild->currentFile() == canonicalFilePath)
            return mdiChild;
    }
    return 0;
}

void MainWindow::RFCLoad( uint32_t iRFCNum )
{
  QString qFilename=QString("rfc%1.txt").arg(iRFCNum);
  QFileInfo qFileInfo;

  // Check if the required RFC is not yet opened
  for (int i=0;i<m_qTabWidget->count();i++)
  {
      MdiChild *mdiChild = qobject_cast<MdiChild *>(m_qTabWidget->widget(i));
      qFileInfo.setFile(mdiChild->currentFile());
      if (qFileInfo.fileName() == qFilename)
      {
          m_qTabWidget->setCurrentWidget(mdiChild);
          return;
      }
  }

  m_pRFCLoader->GetFile(iRFCNum);
}

void MainWindow::RFCStart(const QString &qFilename)
{ // Start downloading a RFC file
  //statusBar()->showMessage(tr("Downloading ")+qFilename);
  statusBar()->clearMessage();
  statusBar()->addWidget(m_pProgressBar);
  m_pProgressBar->show();
}

void MainWindow::RFCReady(const QString &qFilename)
{
  QFileInfo qFileInfo(qFilename);
  MdiChild *child = createMdiChild(qFileInfo.fileName());
  child->setCurrentFont(m_qFont);
  if (child->loadFile(qFilename)) {
      child->show();
  } else {
      child->close();
  }
  statusBar()->showMessage(tr("Load ")+qFilename , 2000);
  statusBar()->removeWidget(m_pProgressBar);
}

void MainWindow::updateRFCProgress(int bytesRead, int totalBytes)
{
    m_pProgressBar->setMaximum(totalBytes);
    m_pProgressBar->setValue(bytesRead);
}
